src/hotspot/share/gc/z/zNMT.cpp

```diff
@@ -28,92 +28,28 @@
-  assert(_num_reservations > 0, "at least one reservation must exist");
-
-  size_t index = 0;
-  *offset_in_reservation = untype(offset);
-  for (; index < _num_reservations; ++index) {
-    const size_t reservation_size = _reservations[index]._size;
-    if (*offset_in_reservation < reservation_size) {
-      break;
-    }
-    *offset_in_reservation -= reservation_size;
-  }
-
-  assert(index != _num_reservations, "failed to find reservation index");
-  return index;
+void ZNMT::reserve(zaddress_unsafe start, size_t size) {
+  MemTracker::record_virtual_memory_reserve((address)start, size, CALLER_PC, mtJavaHeap);
 }
-
-void ZNMT::process_fake_mapping(zoffset offset, size_t size, bool commit) {
-  // In order to satisfy NTM's requirement of an 1:1 mapping between committed
-  // and reserved addresses, a fake mapping from the offset into the reservation
-  // is used.
-  //
-  // These mappings from
-  //   [offset, offset + size) -> {[virtual address range], ...}
-  // are stable after the heap has been reserved. No commits proceed any
-  // reservations. Committing and uncommitting the same [offset, offset + size)
-  // range will result in same virtual memory ranges.
-
-  size_t left_to_process = size;
-  size_t offset_in_reservation;
-  for (size_t i = reservation_index(offset, &offset_in_reservation); i < _num_reservations; ++i) {
-    const zaddress_unsafe reservation_start = _reservations[i]._start;
-    const size_t reservation_size = _reservations[i]._size;
-    const size_t sub_range_size = MIN2(left_to_process, reservation_size - offset_in_reservation);
-    const uintptr_t sub_range_addr = untype(reservation_start) + offset_in_reservation;
-
-    // commit / uncommit memory
-    if (commit) {
-      MemTracker::record_virtual_memory_commit((void*)sub_range_addr, sub_range_size, CALLER_PC);
-    } else {
-      ThreadCritical tc;
-      MemTracker::record_virtual_memory_uncommit((address)sub_range_addr, sub_range_size);
-    }
-
-    left_to_process -= sub_range_size;
-    if (left_to_process == 0) {
-      // Processed all nmt registrations
-      return;
-    }
-
-    offset_in_reservation = 0;
-  }
-
-  assert(left_to_process == 0, "everything was not commited");
+void ZNMT::commit(zoffset offset, size_t size) {
+  MemTracker::allocate_memory_in(ZNMT::_device, static_cast<size_t>(offset), size, mtJavaHeap, CALLER_PC);
 }
-
-void ZNMT::reserve(zaddress_unsafe start, size_t size) {
-  assert(_num_reservations < ZMaxVirtualReservations, "too many reservations");
-  // Keep track of the reservations made in order to create fake mappings
-  // between the reserved and commited memory.
-  // See details in ZNMT::process_fake_mapping
-  _reservations[_num_reservations++] = {start, size};
-
-  MemTracker::record_virtual_memory_reserve((void*)untype(start), size, CALLER_PC, mtJavaHeap);
+void ZNMT::uncommit(zoffset offset, size_t size) {
+  MemTracker::free_memory_in(ZNMT::_device, (size_t)offset, size);
```

Comment:
Didn't this suggestion work? If not, maybe you could at least change it to `(size_t)untype(offset)`. We like to try to avoid direct c-style casts of ZGC's typed values, and instead use the functions that perform extra checks.